File: crates/wyrd/wyrd/examples/workflow_gateway.rs
//! OpenAI gateway workflow example; run with `cargo run -p wyrd --example workflow_gateway --all-features`.

use std::sync::Arc;

use wyrd::agent::{OpenAiChatOptions, ProviderRegistry, openai_chat};
use wyrd::{Agent, Workflow};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let gateway_url = std::env::var("GATEWAY_BASE_URL")
        .or_else(|_| std::env::var("OPENAI_BASE_URL"))
        .unwrap_or_else(|_| "https://api.openai.com/v1".to_owned());

    let prompt = openai_chat(
        "gpt-4o-mini",
        OpenAiChatOptions {
            system: Some("You are a helpful assistant.".to_owned()),
            messages: vec!["Summarise: {{topic}}".to_owned()],
            variables: vec!["topic".to_owned()],
            ..OpenAiChatOptions::default()
        },
    )?;
    let gateway_registry = Arc::new(ProviderRegistry::for_provider(
        &prompt.native().request.provider(),
        gateway_url,
        None::<String>,
    )?);

    let researcher = Agent::new(prompt.clone())
        .name("researcher")
        .version("0.1.0")
        .with_provider_registry(gateway_registry);
    let writer = Agent::new(prompt).name("writer").version("0.1.0");

    let workflow =
        Workflow::sequential("gateway-demo", [researcher, writer])?.with_version("0.1.0");
    let run = workflow.run("renewable energy").await?;

    println!("steps completed: {}", run.tasks.len());
    if let Some(text) = run
        .result()
        .and_then(|outcome| outcome.result.as_ref())
        .and_then(|result| result.adapter().text())
    {
        println!("final output: {text}");
    }

    Ok(())
}

File: examples/python/workflow_gateway.py
"""OpenAI gateway workflow example; run with `uv run python examples/python/workflow_gateway.py`."""

from __future__ import annotations

import os

from wyrd import Agent, Prompt, Workflow


def main() -> None:
    gateway_url = (
        os.environ.get("GATEWAY_BASE_URL")
        or os.environ.get("OPENAI_BASE_URL")
        or "https://api.openai.com/v1"
    )
    prompt = Prompt.openai_chat(
        "gpt-4o-mini",
        system="You are a helpful assistant.",
        messages=["Summarise: {{topic}}"],
        variables=["topic"],
    )

    researcher = Agent(
        name="researcher",
        version="0.1.0",
        prompt=prompt,
        provider_base_url=gateway_url,
    )
    writer = Agent(
        name="writer",
        version="0.1.0",
        prompt=prompt,
    )

    workflow = Workflow.sequential("gateway-demo", researcher, writer)
    workflow.set_version("0.1.0")
    run = workflow.run("renewable energy")

    print(f"steps completed: {len(run.outcomes)}")
    if run.final_step_id:
        print(f"final step status: {run.outcomes[run.final_step_id].status}")


if __name__ == "__main__":
    main()
